akka-actor-typed/src/main/scala/akka/actor/typed/internal/WithMdcBehaviorInterceptor.scala

```diff
@@ -73,32 +74,35 @@ import scala.reflect.ClassTag
   }
 
   override def aroundReceive(ctx: TypedActorContext[T], msg: T, target: ReceiveTarget[T]): Behavior[T] = {
+    // FIXME #26537 performance improvement could be to loop over the entries in staticMdc and mdcForMessage
+    // and directly put them on the MDC.getMDCAdapter
     val mdc = merge(staticMdc, mdcForMessage(msg))
-    ctx.asScala.log.asInstanceOf[AbstractLogger].mdc = mdc
+    MDC.getMDCAdapter.setContextMap(mdc.asJava)
     val next =
       try {
         target(ctx, msg)
       } finally {
-        ctx.asScala.log.asInstanceOf[AbstractLogger].mdc = Map.empty
+        MDC.clear()
       }
     next
   }
 
   override def aroundSignal(ctx: TypedActorContext[T], signal: Signal, target: SignalTarget[T]): Behavior[T] = {
-    ctx.asScala.log.asInstanceOf[AbstractLogger].mdc = staticMdc
+    // FIXME #26537 performance improvement, see aroundReceive
+    MDC.getMDCAdapter.setContextMap(staticMdc.asJava)
     try {
       target(ctx, signal)
     } finally {
-      ctx.asScala.log.asInstanceOf[AbstractLogger].mdc = Map.empty
+      MDC.clear()
     }
   }
 
-  private def merge(staticMdc: Map[String, Any], mdcForMessage: Map[String, Any]): Map[String, Any] = {
+  private def merge(staticMdc: Map[String, String], mdcForMessage: Map[String, String]): Map[String, String] = {
     if (staticMdc.isEmpty) mdcForMessage
     else if (mdcForMessage.isEmpty) staticMdc
-    else if (staticMdc.isInstanceOf[HashMap[String, Any]] && mdcForMessage.isInstanceOf[HashMap[String, Any]]) {
+    else if (staticMdc.isInstanceOf[HashMap[String, String]] && mdcForMessage.isInstanceOf[HashMap[String, String]]) {
       // merged is more efficient than ++
-      mdcForMessage.asInstanceOf[HashMap[String, Any]].merged(staticMdc.asInstanceOf[HashMap[String, Any]])(null)
+      mdcForMessage.asInstanceOf[HashMap[String, String]].merged(staticMdc.asInstanceOf[HashMap[String, String]])(null)
```

Comment:
How do we know for sure it is a HashMap, wasn't it given to us by the user?

Reply:
This is some optimization to prefer usage of `merged` for that case (was like this since before). I have a much better idea for optimizing this, so it will be removed.

Reply:
fixed in https://github.com/akka/akka/pull/27610